lib/Extension/LanguageServerPhpCsFixer/Provider/PhpCsFixerDiagnosticsProvider.php

```diff
@@ -0,0 +1,230 @@
+                                : new Position($lineNo, 0);
+
+                            if ($replacedLines) {
+                                $firstLineA = $replacedLines[0]->getContent();
+                                $firstLineB = $changedLines[0]->getContent();
+                                $lastLineA = end($replacedLines)->getContent();
+                                $lastLineB = end($changedLines)->getContent();
+
+                                $startChars = StringSharedChars::startLength($firstLineA, $firstLineB);
+                                $endChars = StringSharedChars::endPos($lastLineA, $lastLineB);
+
+                                $startPos = new Position($startLineNo, $startChars);
+                                $endPos = new Position($lineNo - 1, $endChars);
+                            }
+
+                            $diagnostics[] = Diagnostic::fromArray([
+                                'message' => yield $this->explainRule($rule),
+                                'range' => new Range($startPos, $endPos),
+                                'severity' => DiagnosticSeverity::WARNING,
+                                'source' => $this->name().':'.$rule
+                            ]);
+
+                            $startLineNo = null;
+                            $changedLines = [];
+                        }
+
+                        if ($line->getType() === Line::UNCHANGED) {
+                            continue;
+                        }
+
+                        if ($line->getType() === Line::REMOVED) {
+                            $startLineNo = $lineNo;
+                            $changedLines[] = $line;
+
+                            continue;
+                        }
+
+                        // new line additions
+                        if ($line->getType() === Line::ADDED && $chunk->getLines()[$index - 1]->getType() === Line::UNCHANGED) {
+                            $diagnostics[] = Diagnostic::fromArray([
+                                'message' => yield $this->explainRule($rule),
+                                'range' => new Range(new Position($lineNo, 0), new Position($lineNo, 1)),
+                                'severity' => DiagnosticSeverity::WARNING,
+                                'source' => $this->name().':'.$rule
+                            ]);
+
+                            $contextLines = [];
+
+                            continue;
+                        }
+
+                        throw new LogicException('This is not intended to be reached');
+                    }
+                }
+            }
+
+            return $diagnostics;
+        });
+    }
+
+    public function provideActionsFor(TextDocumentItem $textDocument, Range $range, CancellationToken $cancel): Promise
+    {
+        return \Amp\call(function () use ($textDocument, $cancel) {
+            $title = 'Format with PHP CS Fixer';
```

Comment:
I'm not sure this should be a code action. There is it's (already?) available as the document formatting request: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_formatting

Reply:
I think it was driven by the editor in my scenario.

For diagnostics, my editor is able to provide quickfix link that show just below the diagnostic content, which seems a bit convenient.
However, thinking further, that link would fix entire file which may get confusing why other problems are fixed, while user may have expected to fix only that specific problem.

I also revieved other LSP servers, and some of them do provide quickfix, and as expected that quickfix resolves only that specific problem.

Therefore I've removed it (well replaced with `source.fixAll` command name, more on that within other comment)